inject/src/main/java/io/micronaut/context/AnnotationReflectionUtils.java

```diff
@@ -0,0 +1,239 @@
+        for (int i = 0; i < annotatedInterfaces.length; i++) {
+            AnnotatedType annotatedInterface = annotatedInterfaces[i];
+            Class<?> classFromType = getClassFromType(annotatedInterface.getType());
+            Argument<?> resolvedImplementation;
+            if (implementedType.equals(classFromType)) {
+                resolvedImplementation = findResolvedImplementation(implementationType, implementedType, annotatedInterface, resolvedVariables);
+            } else {
+                resolvedImplementation = findImplemented(interfaces[i], implementedType, resolvedVariables(annotatedInterface, interfaces[0]));
+            }
+            if (resolvedImplementation != null) {
+                return resolvedImplementation;
+            }
+        }
+        return null;
+    }
+
+    private static Class<?> getClassFromType(Type type) {
+        if (type instanceof Class<?> classType) {
+            return classType;
+        }
+        if (type instanceof ParameterizedType parameterizedType) {
+            return getClassFromType(parameterizedType.getRawType());
+        }
+        if (type instanceof GenericArrayType) {
+            return Object[].class;
+        }
+        if (type instanceof WildcardType wildcardType) {
+            return getClassFromType(wildcardType.getUpperBounds()[0]);
+        }
+        throw new IllegalArgumentException("Unknown type: " + type);
+    }
+
+    private static Map<String, AnnotatedType> resolvedVariables(AnnotatedType annotatedType, Class<?> type) {
+        if (annotatedType instanceof AnnotatedParameterizedType parameterizedType) {
+            AnnotatedType[] actualTypeArguments = parameterizedType.getAnnotatedActualTypeArguments();
+            TypeVariable<? extends Class<?>>[] typeParameters = type.getTypeParameters();
+            if (actualTypeArguments.length != typeParameters.length) {
+                return Map.of();
+            }
+            Map<String, AnnotatedType> resolvedVariables = new LinkedHashMap<>();
+            for (int i = 0; i < actualTypeArguments.length; i++) {
+                AnnotatedType actualTypeArgument = actualTypeArguments[i];
+                if (actualTypeArgument.getType() instanceof TypeVariable<?>) {
+                    // Avoid unresolved type variables
+                    continue;
+                }
+                resolvedVariables.put(
+                    typeParameters[i].getName(),
+                    actualTypeArgument
+                );
+            }
+            return resolvedVariables;
+        }
+        return Map.of();
+    }
+
+    @Nullable
+    private static Argument<?> argumentOf(@NonNull AnnotatedType type,
+                                          @NonNull Class<?> interfaceType,
+                                          @NonNull Map<String, AnnotatedType> resolvedVariables) {
+        if (type.getType() instanceof TypeVariable<?> typeVariable) {
```

Comment:
It only resolve the generic type of the implemented class not of the superclasses / interfaces. If you are loocking for that we need to geet it from the resolvd variables.

Reply:
what I mean is that I believe that argumentOf is always called with one of the types that is directly declared inside an extends or implements expression, so it can never be a type variable (you cant `class A<T> extends T`)